### PR TITLE
RxFactor SelectProjectViewModel

### PR DESCRIPTION
**Pull request by arturs105**
## What's this?
This PR is an RxFactor for the `SelectProjectViewModel` and `SelectProjectViewController`, in other words: the iOS and Foundation parts of the RxFactor. The Android parts will come a bit later, but there's no reason, why these changes can't be reviewed now.

### Relationships
Ref #4012 

## How is it done?
By using the new table view sources. Also: because this view shares cells and some models with the start time entry view, i created new cells, instead of modifying the old ones, so i wouldn't break the start time entry view. The new cells look exactly the same, but use the Rx bindings instead of mvx. These cells are: `ReactiveWorkspaceHeaderViewCell`, `ReactiveProjectSuggestionViewCell` and `ReactiveTaskSuggestionViewCell`. These will be usable in the RxFactor of the start time entry view as well and once that is done, we will be able to remove the old cells for those entities.

## :squid: Permissions
🦑

### CODE REVIEW

**Comment by heytherewill** [comment]
Will the droid bits come in a separate PR or still in this one?

Also, would you mind updating the description of the Start Time Entry RxFactor so the person knows that these cells need to be removed?

**Comment by arturs105** [comment]
Yes, droid parts will come later. That's why i'm not merging into `develop`. And i've already added a comment in the start time entry RxFactor issue, but i could add more info to that comment.

File: Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs [hunk 1]
```diff
@@ -116,10 +82,29 @@ public string PlaceholderText
...
-            ToggleTaskSuggestionsCommand = new MvxCommand<ProjectSuggestion>(toggleTaskSuggestions);
+            Close = rxActionFactory.FromAsync(close);
+            ToggleTaskSuggestions = rxActionFactory.FromAction<ProjectSuggestion>(toggleTaskSuggestions);
+            SelectProject = rxActionFactory.FromAsync<AutocompleteSuggestion>(selectProject);
+
+            IsEmpty = dataSource.Projects.GetAll().Select(projects => projects.None());
+            PlaceholderText = IsEmpty.Select(isEmpty => isEmpty ? Resources.EnterProject : Resources.AddFilterProjects);
+        }
+
+        private bool shouldSuggestCreation(string text, IEnumerable<AutocompleteSuggestion> suggestions)
```
Comment (Odrakir) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs]: I would consider taking a look at what Simon did for expanding / collapsing the main table view. You won't need the subject anymore. I think it's a cleaner approach, but it's a big change over this, so it might be better to do it in another PR. Your call.

File: Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs [hunk 2]
```diff
@@ -1,93 +1,95 @@
...
-                return taskCellIdentifier;
+            var autocompleteSuggestion = ModelAt(indexPath);
 
-            throw new ArgumentException($"{nameof(item)} must be either of type {nameof(ProjectSuggestion)} or {nameof(TaskSuggestion)}.");
+            switch (autocompleteSuggestion)
+            {
+                case ProjectSuggestion projectSuggestion:
+                    var projectCell = (ReactiveProjectSuggestionViewCell)tableView.DequeueReusableCell(ReactiveProjectSuggestionViewCell.Key, indexPath);
+                    projectCell.Item = projectSuggestion;
+                    projectCell.ToggleTaskSuggestions.Subscribe(toggleTaskSuggestionsSubject);
```
[the remark this answers is not in the document]
Comment (arturs105) [reply Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]: That means that every cell would have a dispose bag. Couldn't the bag be in the table view source?

File: Toggl.Daneel/Views/StartTimeEntry/ReactiveProjectSuggestionViewCell.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+            FadeView.FadeRight = true;
+            ClientNameLabel.LineBreakMode = UILineBreakMode.TailTruncation;
+            ProjectNameLabel.LineBreakMode = UILineBreakMode.TailTruncation;
+            ToggleTasksButton.TouchUpInside += toggleTaskSuggestions;
+        }
+
+        protected override void UpdateView()
+        {
+            if (Item is ProjectSuggestion projectSuggestion)
```
[the remark this answers is not in the document]
Comment (arturs105) [reply Toggl.Daneel/Views/StartTimeEntry/ReactiveProjectSuggestionViewCell.cs]: 🤔  Yeah, actually, we can just use a cast, since the `else` branch of this `if` throws anyway. We'd get a bit different exception, but it would still be equally clear, what went wrong there.

File: Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Odrakir) [reply Toggl.Daneel/ViewSources/SelectProjectTableViewSource.cs]: No because you want the subscription to be disposed when the cell recycles.

File: Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs (hunk 1 above)
Comment (arturs105) [reply Toggl.Foundation.MvvmCross/ViewModels/SelectProjectViewModel.cs]: Haven't yet checked that code, so i'd rather do it in a separate PR. Also: you commented about expanding on code that deals with suggesting creation, did you mean to comment on task expanding/collapsing?

**Comment by arturs105** [comment]
Closing this, since `des/rx-select-project` branch contains cherry picked commits from this branch and thus is a cleaner version of this.
